### PR TITLE
Avro: Change union read schema from hive to trino

### PR DESCRIPTION
**Pull request by rzhang10**
This rb changes the avro spark reader code path to read a union data type of `[int, string]` to a spark schema/data of `[tag, field0, field1]` instead of the previous `[tag_0, tag_1]`.
it also fixed an issue that when the avro union is a null, the reader should return a plain `null` instead of `[0, null, null]` or `[null, null]` (as the previous schema did).

### CODE REVIEW

**Comment by funcheetah** [comment]
Thanks @rzhang10 for the PR! Could you please integrate test to make sure the new schema works?

File: core/src/test/java/org/apache/iceberg/avro/TestUnionSchemaConversions.java [hunk 1]
```diff
@@ -108,22 +92,4 @@ public void testOptionSchema() {
 
     Assert.assertEquals(expectedIcebergSchema, icebergSchema.toString());
   }
-
```
Comment (autumnust) [inline core/src/test/java/org/apache/iceberg/avro/TestUnionSchemaConversions.java:111]: is this considered invalid at all? or what's the reason to get rid of this case ?

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java [hunk 2]
```diff
@@ -292,7 +292,7 @@ protected void set(InternalRow struct, int pos, Object value) {
     }
   }
 
-  static class UnionReader implements ValueReader<InternalRow> {
+  private static class UnionReader implements ValueReader<Object> {
```
Comment (autumnust) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: is this modifier change necessary ? this might make it harder for us to contributing back in the future.

File: core/src/main/java/org/apache/iceberg/avro/SchemaToType.java [hunk 3]
```diff
@@ -116,12 +117,13 @@ public Type union(Schema union, List<Type> options) {
       }
     } else {
       // Complex union
```
Comment (autumnust) [inline core/src/main/java/org/apache/iceberg/avro/SchemaToType.java:119]: is there an official definition of what is complex and what is simple ? If we target to contribute it back, maybe better to clarify that thru javadoc -- If this is merely internally usage I am fine with what we have now.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java [hunk 4]
```diff
@@ -316,20 +316,30 @@ public InternalRow read(Decoder decoder, Object reuse) throws IOException {
...
+      Object value = readers[index].read(decoder, reuse);
 
       if (nullIndex < 0) {
-        struct.update(index, value);
+        struct.update(index + 1, value);
+        struct.setInt(0, index);
       } else if (index < nullIndex) {
+        struct.update(index + 1, value);
+        struct.setInt(0, index);
+      } else {
```
Comment (autumnust) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java:340]: Not sure if I follow this, what does the relative position between value index and null index have anything to do with how we assign the value in the struct ?

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 2 above)
Comment (autumnust) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: do you also need to do the same for the vectorized orc reader?

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java [hunk 5]
```diff
@@ -305,7 +305,7 @@ private UnionReader(Schema schema, List<ValueReader<?>> readers) {
     }
 
     @Override
-    public InternalRow read(Decoder decoder, Object reuse) throws IOException {
+    public Object read(Decoder decoder, Object reuse) throws IOException {
```
Comment (autumnust) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: why is this needed if you are still returning `struct` which is a `InternalRow` ?

**Review by autumnust — CHANGES_REQUESTED** [review]

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 4 above)
Comment (rzhang10) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java:340]: Because in avro the nullability of a union type is presented by the existence of a `NULL` type inside the union alternatives, and the `NULL` type can occur in any order inside the union, but in our converted struct, we don't have a field that corresponds to that `NULL` type, thus we have `-1` field count.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 5 above)
Comment (rzhang10) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: See line 323, if it's null data, we will return `null` directly, so this method returns either `null` or an `InternalRow`, so the return type should only be `Object`. Please see the same method signature in `ValueReaders`.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 2 above)
Comment (rzhang10) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: This mimics the same access control modifier for the similar method in this class `SparkValueReaders`, the relevant change has been made in `SparkOrcValueReaders` in pr #85 .

File: core/src/test/java/org/apache/iceberg/avro/TestUnionSchemaConversions.java (hunk 1 above)
Comment (rzhang10) [reply core/src/test/java/org/apache/iceberg/avro/TestUnionSchemaConversions.java:111]: Yes, this schema itself shouldn't appear in the first place, user shouldn't define this kind of schema.

File: core/src/main/java/org/apache/iceberg/avro/SchemaToType.java (hunk 3 above)
Comment (rzhang10) [reply core/src/main/java/org/apache/iceberg/avro/SchemaToType.java:119]: simple union is `[sometype, null]` while complex union is `[sometype1, sometype2, ...]`, where there are at least 2 non-null types in the union, I think we can probably add a java doc in `AvroSchemaUtil.isOptionSchema`, which is an existing method in upstream iceberg.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 5 above)
Comment (autumnust) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: So why not returning null directly at line 323 if `nullIndex` is hit ?

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 4 above)
Comment (autumnust) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java:340]: I guess what I didn't know earlier was, if NULL type appears in the union it seems it has to be the first position in 
the alternative list ? But the avro spec said "Thus, for unions containing "null", the "null" is usually listed first," 

Will this lead to a silent failure if you don't check this first?

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 2 above)
Comment (autumnust) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: What about the modifier change ? Is that really needed?

**Comment by autumnust** [comment]
A couple of more comments ... @rzhang10

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 5 above)
Comment (rzhang10) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: `readers[index].read(decoder, reuse)` will return `null`, I just feel writing this way will make the code more consistent with later part.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 4 above)
Comment (rzhang10) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java:340]: I think that's just a recommendation from the Avro spec, not mandated, in fact, in our ecosystem there are many avro schemas which failed to put `null` as the first type element in the union.

And that's why I'm specifically computing the null index here and branching on it.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 2 above)
Comment (rzhang10) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: Yeah if you look at other nested class in this file you will see:
`private static class MapReader`
`private static class EnumReader`
`private static class ArrayReader`

so I think it's best to make the modifiers consistent.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 5 above)
Comment (autumnust) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: Hmm not sure that's worthwhile with the cost of changing the method signature ... But I don't have strong opinion on this either.

**Review by autumnust — APPROVED** [review]
LGTM. Most of conservations above are just questions. I still believe we shouldn't change the signature of union read method from `InternalRow` to `Object` since a return type as `Object` doesn't seem to right if we can make it more specific. But I am no strong opinion on that and don't take that as blocker.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java (hunk 5 above)
Comment (rzhang10) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java]: I think using `InternalRow` is better, let me make the change accordingly. Thanks for spotting this!
